File: dp/0002_lcs/lcs.go
package lcs

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func lcs(dp [][]int, p, q string, n, m int) int {
	if n == 0 || m == 0 {
		return 0
	}

	if dp[n][m] != -1 {
		return dp[n][m]
	}

	var result int
	if p[n-1] == q[m-1] {
		result = 1 + lcs(dp, p, q, n-1, m-1)
	} else {
		tmp1 := lcs(dp, p, q, n-1, m)
		tmp2 := lcs(dp, p, q, n, m-1)
		result = max(tmp1, tmp2)
	}
	dp[n][m] = result
	return result
}

func lcs_dp(p, q string) int {
	n := len(p)
	m := len(q)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, m+1)
		for j := range dp[i] {
			dp[i][j] = -1
		}
	}
	return lcs(dp, p, q, n, m)
}

func lcs_bottom_up(p, q string) int {
	n := len(p)
	m := len(q)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, m+1)
	}

	for i := 0; i <= n; i++ {
		for j := 0; j <= m; j++ {
			if i == 0 || j == 0 {
				dp[i][j] = 0
			} else if p[i-1] == q[j-1] {
				dp[i][j] = 1 + dp[i-1][j-1]
			} else {
				dp[i][j] = max(dp[i-1][j], dp[i][j-1])
			}
		}
	}
	return dp[n][m]
}

File: dp/0002_lcs/lcs.py
#!/usr/bin/python

import unittest

def max(a, b):
    return a if a >= b else b

def lcs(P, Q, m, n, memo):
    if m == 0 or n == 0:
        return 0

    if memo[m-1][n-1] != -1:
        return memo[m-1][n-1]

    if P[m-1] == Q[n-1]:
        memo[m-1][n-1] = 1 + lcs(P, Q, m-1, n-1, memo)
    else:
        tmp1 = lcs(P, Q, m-1, n, memo)
        tmp2 = lcs(P, Q, m, n-1, memo)
        memo[m-1][n-1] = max(tmp1, tmp2)

    return memo[m-1][n-1]

def lcs_main(P, Q):
    m, n = len(P), len(Q)
    memo = [[-1 for _ in range(n)] for _ in range(m)]
    return lcs(P, Q, m, n, memo)

class TestLCS(unittest.TestCase):
    def test_lcs(self):
        self.assertEqual(lcs_main("ABCBDAB", "BDCABC"), 4)  # LCS is "BCAB" or "BDAB"
        self.assertEqual(lcs_main("", ""), 0)
        self.assertEqual(lcs_main("AXYT", "AYZX"), 2)  # LCS is "AY"
        self.assertEqual(lcs_main("12345", "54321"), 1)  # LCS could be any single digit
        self.assertEqual(lcs_main("AAAA", "AAAA"), 4)  # LCS is "AAAA"

if __name__ == "__main__":
    unittest.main()
